modules/crypt.cpp

```diff
@@ -43,16 +42,145 @@
 #include <znc/Chan.h>
 #include <znc/User.h>
 #include <znc/IRCNetwork.h>
+#include <openssl/dh.h>
+#include <openssl/bn.h>
+#include <znc/SHA256.h>
 
 #define REQUIRESSL 1
 #define NICK_PREFIX_KEY "[nick-prefix]"
 
 class CCryptMod : public CModule {
+  private:
+    /*
+     * As used in other implementations like KVIrc, fish10, Quassel, FiSH-irssi, ...
+     * all the way back to the original located at http://mircryption.sourceforge.net/Extras/McpsFishDH.zip
+     */
+    const char* prime1080 = "FBE1022E23D213E8ACFA9AE8B9DFADA3EA6B7AC7A7B7E95AB5EB2DF858921FEADE95E6AC7BE7DE6ADBAB8A783E7AF7A7FA6A2B7BEB1E72EAE2B72F9FA2BFB2A2EFBEFAC868BADB3E828FA8BADFADA3E4CC1BE7E8AFE85E9698A783EB68FA07A77AB6AD7BEB618ACF9CA2897EB28A6189EFA07AB99A8A7FA9AE299EFA7BA66DEAFEFBEFBF0B7D8B";
+    /* Generate our keys once and reuse, just like ssh keys */
+    DH* m_pDH = nullptr;
+    CString m_sPrivKey;
+    CString m_sPubKey;
+
+#if OPENSSL_VERSION_NUMBER < 0X10100000L
+    static int DH_set0_pqg(DH* dh, BIGNUM* p, BIGNUM* q, BIGNUM* g) {
+        /* If the fields p and g in dh are nullptr, the corresponding input
+         * parameters MUST be non-nullptr.  q may remain nullptr.
+         */
+        if (dh == nullptr || (dh->p == nullptr && p == nullptr) || (dh->g == nullptr && g == nullptr))
+            return 0;
+
+        if (p != nullptr) {
+            BN_free(dh->p);
+            dh->p = p;
+        }
+        if (g != nullptr) {
+            BN_free(dh->g);
+            dh->g = g;
+        }
+        if (q != nullptr) {
+            BN_free(dh->q);
+            dh->q = q;
+            dh->length = BN_num_bits(q);
+        }
+
+        return 1;
+    }
+
+    static void DH_get0_key(const DH* dh, const BIGNUM** pub_key, const BIGNUM** priv_key) {
+        if (dh != nullptr) {
+            if (pub_key != nullptr)
+                *pub_key = dh->pub_key;
+            if (priv_key != nullptr)
+                *priv_key = dh->priv_key;
+        }
+    }
+
+#endif
+
+    bool DH1080_gen() {
+        /* Generate our keys on first call */
+        if (!m_pDH) {
+            int len;
+            const BIGNUM* bPrivKey = nullptr;
+            const BIGNUM* bPubKey = nullptr;
+            BIGNUM* bPrime = nullptr;
+            BIGNUM* bGen = nullptr;
+            m_pDH = DH_new();
+
+            if (!BN_hex2bn(&bPrime, prime1080) || !BN_dec2bn(&bGen, "2") || !DH_set0_pqg(m_pDH, bPrime, nullptr, bGen) || !DH_generate_key(m_pDH)) {
+                /* one of them failed */
+                if (bPrime)
+                    BN_clear_free(bPrime);
+                if (bGen)
+                    BN_clear_free(bGen);
+                if (m_pDH) {
+                    DH_free(m_pDH);
+                    m_pDH = nullptr;
```

Comment:
Will std::unique_ptr handle this correctly? m_pDH needs to be put back to nullptr if any of those commands failed.